File: tests/grpc/client_tls.nim
{.define: grpcTls.}
import ../../src/nimproto3

importProto3 currentSourcePath.parentDir & "/test_service.proto"

# =============================================================================
# MAIN CLIENT TEST
# =============================================================================

when isMainModule:
  proc runTests() {.async.} =
    echo "================================================================================"
    echo "Nim gRPC Client with TLS - Local Test"
    echo "================================================================================"

    # Create client with TLS
    let client = newGrpcClient("localhost", 50051, CompressionIdentity, sslVerify = false)
    
    echo "[Client] Connecting to server..."
    await client.connect()
    echo "[Client] Connected, waiting for settings exchange..."
    await sleepAsync(200) # Wait for settings exchange
    echo "[Client] Settings exchange complete"

    echo "\n[Test 1] SimpleTest (Unary)"
    try:
      echo "[Client] Sending SimpleTest request..."
      let reply = await client.simpleTest(
        TestRequest(message: "Hello from Nim TLS client!", counter: 42)
      )
      echo "Response: response='", reply.response, "', received=", reply.received
    except:
      echo "Error: ", getCurrentExceptionMsg()
      echo "Stack trace:"
      echo getCurrentException().getStackTrace()
    
    echo "\n[Test 2] StreamTest (stream)"
    try:
      echo "[Client] Sending StreamTest request..."
      await client.streamTest(
        @[TestRequest(message: "Hello from Nim TLS client!", counter: 37)]
      )
      let response = await client.streamTestGetResponse()
      echo "Response: response='", response
      let addition_response = await client.streamTestGetResponse()
      echo "Response: addition_response='", addition_response
    except:
      echo "Error: ", getCurrentExceptionMsg()
      echo "Stack trace:"
      echo getCurrentException().getStackTrace()

    echo "\n[Test 3] Second StreamTest (stream)"
    try:
      echo "[Client] Sending StreamTest request..."
      await client.streamTest(
        @[TestRequest(message: "Hello from Nim TLS client!", counter: 37)]
      )
      let response = await client.streamTestGetResponse()
      echo "Response: response='", response
      let additional_response = await client.streamTestGetResponse()
      echo "Response: additional_response='", additional_response
      await client.streamTestCloseStream()
    except:
      echo "Error: ", getCurrentExceptionMsg()
      echo "Stack trace:"
      echo getCurrentException().getStackTrace()

    client.close()
    echo "[Client] Connection closed"

    echo "\n================================================================================"
    echo "All tests completed successfully!"
    echo "================================================================================"
  
  waitFor runTests()

File: tests/grpc/client_tls.py
#!/usr/bin/env python3
"""
gRPC Client with TLS for testing
"""
import grpc
import sys
import os

# Import generated protobuf files
import test_service_pb2
import test_service_pb2_grpc

def run():
    # Read server certificate
    cert_dir = os.path.dirname(os.path.abspath(__file__))
    with open(os.path.join(cert_dir, 'server.crt'), 'rb') as f:
        cert = f.read()
    
    # Create credentials
    credentials = grpc.ssl_channel_credentials(root_certificates=cert)
    
    # Create channel with TLS
    with grpc.secure_channel('localhost:50051', credentials) as channel:
        stub = test_service_pb2_grpc.TestServiceStub(channel)
        
        # Test 1: Simple unary call
        print("=" * 80)
        print("Test 1: SimpleTest (Unary)")
        print("=" * 80)
        request = test_service_pb2.TestRequest(message="Hello from Python TLS client!", counter=42)
        response = stub.SimpleTest(request)
        print(f"Response: response='{response.response}', received={response.received}")
        
        # Test 2: Bidirectional streaming
        print("\n" + "=" * 80)
        print("Test 2: StreamTest (Bidirectional)")
        print("=" * 80)
        
        def request_generator():
            for i in range(3):
                yield test_service_pb2.TestRequest(message=f"Message {i+1}", counter=i+1)
        
        responses = stub.StreamTest(request_generator())
        for i, response in enumerate(responses):
            print(f"Response {i+1}: response='{response.response}', received={response.received}")
        
        print("\n" + "=" * 80)
        print("All tests completed successfully!")
        print("=" * 80)

if __name__ == '__main__':
    run()
